Add Open Telekom Cloud KMS wrapper

mrclki opened the pull request:
This PR introduces a new KMS wrapper to support Open Telekom Cloud.

This wrapper enables auto-unseal functionality for Open Telekom Cloud's KMS service and is a prerequisite for a forthcoming openbao PR that will add the necessary seal configuration stanza.

mrclki committed:
Add Open Telekom Cloud KMS wrapper implementation

Signed-off-by: Marcel Kirschner <marcel.kirschner@peertech.de>

sfhl commented:
any update on this? we would like to see this integrated as well

cipherboy commented:
Hey @sfhl and @mrclki! Now that externalized KMS wrappers have been added (for v2.6.0), I think this can be rebased and can proceed. I think @satoqz is working on documentation to better document the KMS plugin creation process. 

Is anyone interested in sponsoring access to Open Telekom Cloud KMS keys for the project?

satoqz commented:
> I think @satoqz is working on documentation to better document the KMS plugin creation process.

I primarily work on user-facing docs, from a developer PoV not much has changed. Just upstream an implementation of the `Wrapper` interface as defined in https://github.com/openbao/go-kms-wrapping/blob/main/wrapper.go (this now has new and improved docstrings to help authors better understand what is expected!) and add an entrypoint at https://github.com/openbao/openbao-plugins/tree/main/kms, then you're golden. (But I think @mrclki already knows this, just noting it here 😄)

> Is anyone interested in sponsoring access to Open Telekom Cloud KMS keys for the project?

@cipherboy FWIW I don't think we should block contributions until this is in place, should we? Given it's an easily patchable plugin, I'm fine accepting mocked tests to start with and a confirmation that functionality has been manually tested against OpenBao. Contributing live tests out of the box is a big plus nevertheless.

mrclki commented:
@cipherboy As discussed, I will continue to fulfill the PR and help with the sponsoring of the tenant.

cipherboy commented:
@satoqz said:

> @cipherboy FWIW I don't think we should block contributions until this is in place, should we? Given it's an easily patchable plugin, I'm fine accepting mocked tests to start with and a confirmation that functionality has been manually tested against OpenBao. Contributing live tests out of the box is a big plus nevertheless.

No, I don't think so, but I'd also like to put the project in a good shape to have reliable review/maintenance capabilities. If we can ask contributors to sponsor (especially in cases when this is a first-party code contribution of the KMS support -- not that this is one), I think it'd be great to ask proactively.

I do think scope is different, though. For unseal, I don't _really_ care too much about provider-specific quirks. If $PETCLOUD _claims_ to do AES-256 but actually does AES-253 (whatever that is :laughing:) and doesn't interoperate with the rest of the world, it doesn't really matter. Seals don't behave in a way that requires interoperability across providers and we place faith in the provider (anyways) that it is secure.

For external keys (and full KMS support in the future), being able to show integration/compatibility is very useful, IMHO, and having _support_ (from a community/resource/testing PoV) is a little more important.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,293 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
```
I assume this code isn't associated with HashiCorp in any way so we should use the OpenBao copyright headers (check any recently created files for reference, e.g. `kms/kms.go`).

satoqz commented on wrappers/tcloudpublickms/tcloudpublickms.go:
```diff
@@ -0,0 +1,293 @@
+// CMK. This returns the ciphertext, and/or any errors from this call. This
+// should be called after the KMS client has been instantiated.
+func (k *Wrapper) Encrypt(
+	_ context.Context,
```
Does the sdk not allow propagating contexts?

mrclki replied:
Unfortunately no 😢

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,293 @@
+
+	// Check and set KeyId
+	keyId, err := getConfig(
+		"kms_key_id",
```
```suggestion
		"key_id",
```

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,293 @@
+	keyInfo, err := kms.Get(k.client, k.keyId)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching Open Telekom Cloud KMS key information: %w", err)
+	}
```
It's a good idea to defer this entirely to `Encrypt/Decrypt` so Auto Seals can start up without the key existing, and then poll until it's created. Just a minor UX thing, but OpenBao supports this, so it's good not to remove that merit here.

mrclki committed:
Change SPDX header

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,293 @@
+func (k *Wrapper) setupClient(opts *options) error {
+	region, err := getConfig(
+		"region",
+		os.Getenv("OPENTELEKOMCLOUD_REGION"),
```
Want to add constants for these, too?

satoqz commented on wrappers/tcloudpublickms/tcloudpublickms.go:
```diff
@@ -0,0 +1,293 @@
+	keyId string
+
+	// currentKeyId is the resolved key ID from the last successful operation.
+	currentKeyId *atomic.Value
```
Not sure, but can this be a

```suggestion
	currentKeyId atomic.Value
```

mrclki replied:
Well first, all existing wrapper implementations in the repo use a pointer. I'd propose keeping *atomic.Value. It's the more defensive choice and protects against consumer misuse without any downside.

mrclki replied:
It should only matter when someone accidentally copies the struct by value: *atomic.Value stays safe, while atomic.Value leads to a  data race.

satoqz replied:
That's fair, I wish we returned a private type behind `wrapping.Wrapper` but someone made that choice a while ago...

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,293 @@
+)
+
+const (
+	EnvOpenTelekomCloudKmsWrapperKeyId = "OPENTELEKOMCLOUDKMS_WRAPPER_KEY_ID"
```
```suggestion
	EnvOpenTelekomCloudKmsKeyId = "OPENTELEKOMCLOUDKMS_KEY_ID"
```

perhaps.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms_acc_test.go:
```diff
@@ -0,0 +1,111 @@
+func TestOpenTelekomCloudKMS_Lifecycle(t *testing.T) {
+	mustHaveOTCEnv(t)
+
+	ctx := context.Background()
```
```suggestion
	ctx := t.Context()
```

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms_acc_test.go:
```diff
@@ -0,0 +1,111 @@
+
+	// Setting configuration from environment variables.
+	if _, err := w.SetConfig(ctx); err != nil {
+		t.Fatalf("SetConfig failed: %v", err)
```
You can use the `testify/require` package if you like to make these easier to read.

satoqz commented on .github/workflows/go.yml:
```diff
@@ -17,6 +17,7 @@ jobs:
           - './wrappers/huaweicloudkms'
           - './wrappers/kmip'
           - './wrappers/ocikms'
+          - './wrappers/opentelekomcloudkms'
```
I do wonder if there is a shorter name since it is a mouthful, and the longest package name in the repo. T-cloud seems to be an alternative? Then again, the sdk is `opentelekomcloud`...

mrclki replied:
Well some people call it OTC, but that could be open to misinterpretation.

satoqz replied:
"OTC KMS" brings up the right result on search engines, so I'd be happy with `otckms`. We can use the long form display name in docs so it's not ambigous there but a little easier to spell out in code and configuration.

mrclki replied:
To be honest, I don't personally care about the naming. As @sfhl pointed out, they recently rebranded to T Cloud Public. So the abbreviation "TCP" is out... 

Should we go with `tcloudpublickms` which isn't really shorter or `otckms`?

satoqz replied:
Let's go with `tcloudkms` perhaps? Seems like a good idea not to lag behind on the naming. If there's ever a "T-Cloud Private KMS", let's assume it'll follow a similar API so we can integrate it into the same module, and in the worst case create a new one with the longer name.

satoqz replied:
@mrclki and I agreed out-of-band that `tcloudpublickms` will be the best we can do.

mrclki committed:
Change SPDX header

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

mrclki committed:
Fix typo in wrappers/opentelekomcloudkms/options.go

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

mrclki committed:
Refine Open Telekom Cloud KMS wrapper

- Define constants for all environment variable names
- Rename config key from "kms_key_id" to "key_id" for consistency
- Make project (tenant) scope optional (defaults to global KMS endpoint)
- Remove key validation call from SetConfig
- Remove SetConfig_ResolvesKeyId test
- Remove default project value from options
- Use t.Context() instead of context.Background() in tests
- Use testify/require assertions in tests
- Reduce required env vars for acceptance tests (region/project now optional)

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

satoqz commented on wrappers/opentelekomcloudkms/options.go:
```diff
@@ -0,0 +1,145 @@
+	if opts.WithConfigMap != nil {
+		for k, v := range opts.WithConfigMap {
+			switch k {
+			case "kms_id":
```
```suggestion
			case "key_id":
```

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,291 @@
+func (k *Wrapper) setupClient(opts *options) error {
+	region, err := getConfig(
+		"region",
+		os.Getenv(EnvOpenTelekomCloudRegion),
```
These need to take `opts.WithDisallowEnvVars` into account.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,291 @@
+		if err := k.setupClient(opts); err != nil {
+			return nil, err
+		}
+	}
```
```suggestion
	if err := k.setupClient(opts); err != nil {
		return nil, err
	}
```

(or just inline)

I don't think there's a reason to guard against this, seems we get reconfiguration for free even if it is absolutely not required by `Wrapper`'s semantics. I see other, old wrappers do this but it makes no sense.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,291 @@
+)
+
+const (
+	EnvOpenTelekomCloudKmsKeyId         = "OPENTELEKOMCLOUD_KMS_KEY_ID"
```
```suggestion
	EnvOpenTelekomCloudKmsKeyId         = "OPENTELEKOMCLOUD_KEY_ID"
```

I do see how `KMS_KEY_ID` is more specific in this case opposed to the config map, but I wonder if we should simplify nevertheless for consistency.

mrclki replied:
Personally, I would keep it as it is. As you mentioned, it's more specific to the KMS service and its key.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,291 @@
+) ([]byte, error) {
+	if in == nil {
+		return nil, fmt.Errorf("given input for decryption is nil")
+	}
```
This is also something older wrappers do that doesn't really achieve much.

satoqz commented on wrappers/opentelekomcloudkms/opentelekomcloudkms.go:
```diff
@@ -0,0 +1,291 @@
+//
+// Order of precedence Open Telekom Cloud values:
+// * Environment variable
+// * Value from Vault configuration file
```
Isn't the order used below the other way around?

Generally, the established order of precedence to use and document is

1. Environment variables
2. Local options
3. Global options (mainly `WithConfigMap`)

Frankly I dislike that environment variables take the highest precedence... one day I'll compose all the weird parts about `Wrapper` and convince everyone else to break the API.

mrclki committed:
Further refinement

- Respect WithDisallowEnvVars
- Parse "disallow_env_vars" from the config map in options.go
- Replace getConfig helper with explicit switch/case blocks
- Fix "kms_id" typo in config map key
- Remove redundant nil check on Decrypt input
- Remove unnecessary client nil guard in SetConfig

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

satoqz commented on wrappers/opentelekomcloudkms/options.go:
```diff
@@ -0,0 +1,154 @@
+				if err != nil {
+					return nil, err
+				}
+				opts.withDisallowEnvVars = disallowEnvVars
```
You don't need to create a local option for this, just use `WithDisallowEnvVars` as present on the global `wrapping.Options`.

mrclki replied:
Well, I took inspiration from the `awskms` implementation. But by further inspection I see that `gcpckms` uses opts.Options.WithDisallowEnvVars directly from the embedded *wrapping.Options, while awskms has its own local withDisallowEnvVars field and additionally parses "disallow_env_vars" from the ConfigMap.

There seems to be an inconsistency across wrapper implementations here, should there be a general guideline on which approach to follow?

For now, should I follow the gcpckms approach or the awskms approach?

satoqz replied:
Gotcha.. ugh.. Yeah the GCP approach is the correct one. The parameter should also not be controllable via the config map, that undermines the security concept. Seems to be the case for AWS and Azure, and was never documented either.

mrclki committed:
Use global WithDisallowEnvVars option

Use opts.Options.WithDisallowEnvVars from the global wrapping.Options.
Also remove a nil-input decrypt test that passed a nil BlobInfo.

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

sfhl commented:
hi all,
i just wanted to point out that Open Telekom CLoud is renamed to T Cloud Public, so maybe thats something to consider?

mrclki committed:
Rename to T Cloud Public

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

mrclki committed:
Revert quoatation marks in .github/workflows/go.yml

Signed-off-by: Marcel Kirschner <mk@wilaris.de>

cipherboy commented:
Thank you @mrclki! :rocket:
